ClassDef checker error after linking when only using the class data of a non-native JS class

In a project (non-public :( ) that uses [tapir](https://github.com/softwaremill/tapir), compilation with 1.13.0 worked fine, while the upgrade to 1.13.1 throws the following error:

```
[error] file:/home/runner/work/tapir/tapir/serverless/aws/lambda/src/main/scalajs/sttp/tapir/serverless/aws/lambda/js/AwsJsRequest.scala(29:7:ClassDef): JS classes and module classes must have a constructor
[error] java.util.concurrent.ExecutionException: Boxed Error
[error] 	at scala.concurrent.impl.Promise$.resolver(Promise.scala:87)
[error] 	at scala.concurrent.impl.Promise$.scala$concurrent$impl$Promise$$resolveTry(Promise.scala:79)
[error] 	at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
[error] 	at scala.concurrent.Promise.complete(Promise.scala:53)
[error] 	at scala.concurrent.Promise.complete$(Promise.scala:52)
[error] 	at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:187)
[error] 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error] 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
[error] 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[error] 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[error] 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[error] 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[error] 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
[error] Caused by: java.lang.AssertionError: There were 1 ClassDef checking errors after optimizing. Please report this as a bug.
[error] 	at org.scalajs.linker.frontend.Refiner$ClassInfoCache.update(Refiner.scala:161)
[error] 	at org.scalajs.linker.frontend.Refiner$ClassInfoCache.$anonfun$loadInfo$2(Refiner.scala:147)
[error] 	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
[error] 	at scala.util.Success.$anonfun$map$1(Try.scala:255)
[error] 	at scala.util.Success.map(Try.scala:213)
[error] 	at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
[error] 	at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error] 	at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error] 	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
[error] 	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[error] 	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[error] 	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[error] 	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[error] 	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
```

The offending line seems to be this one:
https://github.com/softwaremill/tapir/blob/master/serverless/aws/lambda/src/main/scalajs/sttp/tapir/serverless/aws/lambda/js/AwsJsRequest.scala#L29

No idea how to minimize this, but since it's a public library, it should be possible to reproduce. This bug occurs when compiling tapir 1.2.12, which is also the most recent release, released last week (at the time of writing).

Please let me know if/what further information can help, or to test a possible fix! Really looking forward to those incremental-comp performance upgrades in 1.13.1 :) Thanks!

--- a/ir/shared/src/main/scala/org/scalajs/ir/ClassKind.scala
+++ b/ir/shared/src/main/scala/org/scalajs/ir/ClassKind.scala
@@ -50,6 +50,7 @@ sealed abstract class ClassKind {
     case _                                             => false
   }
 
+  @deprecated("not a meaningful operation", since = "1.13.2")
   def withoutModuleAccessor: ClassKind = this match {
     case ModuleClass         => Class
     case JSModuleClass       => JSClass

--- a/linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analysis.scala
@@ -57,11 +57,11 @@ object Analysis {
     def nonExistent: Boolean
     /** For a Scala class, it is instantiated with a `New`; for a JS class,
      *  its constructor is accessed with a `JSLoadConstructor` or because it
-     *  is needed for a subclass.
+     *  is needed for a subclass. For modules (Scala or JS), the module is
+     *  accessed.
      */
     def isInstantiated: Boolean
     def isAnySubclassInstantiated: Boolean
-    def isModuleAccessed: Boolean
     def areInstanceTestsUsed: Boolean
     def isDataAccessed: Boolean
 

--- a/linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/analyzer/Analyzer.scala
@@ -582,7 +582,7 @@ final class Analyzer(config: CommonPhaseConfig, initial: Boolean,
 
     var isInstantiated: Boolean = false
     var isAnySubclassInstantiated: Boolean = false
-    var isModuleAccessed: Boolean = false
+    private var isModuleAccessed: Boolean = false
     var areInstanceTestsUsed: Boolean = false
     var isDataAccessed: Boolean = false
 

--- a/linker/shared/src/main/scala/org/scalajs/linker/analyzer/InfoLoader.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/analyzer/InfoLoader.scala
@@ -88,7 +88,7 @@ private[analyzer] object InfoLoader {
 
               case InfoLoader.InitialIRCheck =>
                 val errorCount = ClassDefChecker.check(tree,
-                    allowReflectiveProxies = false, allowTransients = false, logger)
+                    postBaseLinker = false, postOptimizer = false, logger)
                 if (errorCount != 0) {
                   throw new LinkingException(
                       s"There were $errorCount ClassDef checking errors.")
@@ -96,7 +96,7 @@ private[analyzer] object InfoLoader {
 
               case InfoLoader.InternalIRCheck =>
                 val errorCount = ClassDefChecker.check(tree,
-                    allowReflectiveProxies = true, allowTransients = true, logger)
+                    postBaseLinker = true, postOptimizer = true, logger)
                 if (errorCount != 0) {
                   throw new LinkingException(
                       s"There were $errorCount ClassDef checking errors after optimizing. " +

--- a/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/ClassEmitter.scala
@@ -903,6 +903,14 @@ private[emitter] final class ClassEmitter(sjsGen: SJSGen) {
          */
         if (kind != ClassKind.JSClass && kind != ClassKind.NativeJSClass) {
           WithGlobals(globalVar("noIsInstance", CoreVar))
+        } else if (kind == ClassKind.JSClass && !globalKnowledge.hasInstances(className)) {
+          /* We need to constant-fold the instance test, to avoid emitting
+           * `x instanceof $a_TheClass()`, because `$a_TheClass` won't be
+           * declared at all. Otherwise, we'd get a `ReferenceError`.
+           */
+          WithGlobals(genArrowFunction(List(js.ParamDef(js.Ident("x"))), None, js.Return {
+            js.BooleanLiteral(false)
+          }))
         } else {
           for {
             jsCtor <- genJSClassConstructor(className, jsNativeLoadSpec,

--- a/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala
@@ -629,9 +629,10 @@ final class Emitter(config: Emitter.Config) {
       }
     }
 
-    if (linkedClass.kind.hasModuleAccessor)
+    if (linkedClass.kind.hasModuleAccessor && linkedClass.hasInstances) {
       addToMain(classTreeCache.moduleAccessor.getOrElseUpdate(
           classEmitter.genModuleAccessor(className, kind)(moduleContext, classCache, linkedClass.pos)))
+    }
 
     // Static fields
 

--- a/linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala
@@ -27,7 +27,7 @@ import org.scalajs.linker.checker.ErrorReporter._
 
 /** Checker for the validity of the IR. */
 private final class ClassDefChecker(classDef: ClassDef,
-    allowReflectiveProxies: Boolean, allowTransients: Boolean, reporter: ErrorReporter) {
+    postBaseLinker: Boolean, postOptimizer: Boolean, reporter: ErrorReporter) {
   import ClassDefChecker._
 
   import reporter.reportError
@@ -102,12 +102,23 @@ private final class ClassDefChecker(classDef: ClassDef,
 
     //// Additional checks
 
-    if (classDef.kind == ClassKind.ModuleClass &&
-        methods(MemberNamespace.Constructor.ordinal).size != 1)
-      reportError("Module class must have exactly 1 constructor")
+    /* After the base linker, we may have DCE'ed away the constructors of
+     * module classes and JS classes that are never instantiated. The classes
+     * may still exist because their class data are accessed.
+     */
+    if (!postBaseLinker) {
+      /* Check that we have exactly 1 constructor in a module class. This goes
+       * together with `checkMethodDef`, which checks that a constructor in a
+       * module class must be 0-arg.
+       */
+      if (classDef.kind == ClassKind.ModuleClass &&
+          methods(MemberNamespace.Constructor.ordinal).size != 1) {
+        reportError("Module class must have exactly 1 constructor")
+      }
 
-    if (classDef.kind.isJSClass && classDef.jsConstructor.isEmpty)
-      reportError("JS classes and module classes must have a constructor")
+      if (classDef.kind.isJSClass && classDef.jsConstructor.isEmpty)
+        reportError("JS classes and module classes must have a constructor")
+    }
   }
 
   private def checkKind()(implicit ctx: ErrorContext): Unit = {
@@ -463,7 +474,7 @@ private final class ClassDefChecker(classDef: ClassDef,
   private def checkMethodNameNamespace(name: MethodName, namespace: MemberNamespace)(
       implicit ctx: ErrorContext): Unit = {
     if (name.isReflectiveProxy) {
-      if (allowReflectiveProxies) {
+      if (postBaseLinker) {
         if (namespace != MemberNamespace.Public)
           reportError("reflective profixes are only allowed in the public namespace")
       } else {
@@ -839,7 +850,7 @@ private final class ClassDefChecker(classDef: ClassDef,
   }
 
   private def checkAllowTransients()(implicit ctx: ErrorContext): Unit = {
-    if (!allowTransients)
+    if (!postOptimizer)
       reportError("invalid transient tree")
   }
 
@@ -897,9 +908,9 @@ object ClassDefChecker {
    *
    *  @return Count of IR checking errors (0 in case of success)
    */
-  def check(classDef: ClassDef, allowReflectiveProxies: Boolean, allowTransients: Boolean, logger: Logger): Int = {
+  def check(classDef: ClassDef, postBaseLinker: Boolean, postOptimizer: Boolean, logger: Logger): Int = {
     val reporter = new LoggerErrorReporter(logger)
-    new ClassDefChecker(classDef, allowReflectiveProxies, allowTransients, reporter).checkClassDef()
+    new ClassDefChecker(classDef, postBaseLinker, postOptimizer, reporter).checkClassDef()
     reporter.errorCount
   }
 

--- a/linker/shared/src/main/scala/org/scalajs/linker/frontend/BaseLinker.scala
+++ b/linker/shared/src/main/scala/org/scalajs/linker/frontend/BaseLinker.scala
@@ -147,15 +147,11 @@ private[frontend] object BaseLinker {
 
     val allMethods = methods ++ syntheticMethodDefs
 
-    val kind =
-      if (classInfo.isModuleAccessed) classDef.kind
-      else classDef.kind.withoutModuleAccessor
-
     val ancestors = classInfo.ancestors.map(_.className)
 
     val linkedClass = new LinkedClass(
         classDef.name,
-        kind,
+        classDef.kind,
         classDef.jsClassCaptures,
         classDef.superClass,
         classDef.interfaces,

--- a/test-suite/js/src/test/scala/org/scalajs/testsuite/compiler/ReflectionTest.scala
+++ b/test-suite/js/src/test/scala/org/scalajs/testsuite/compiler/ReflectionTest.scala
@@ -19,6 +19,8 @@ import org.junit.Test
 import org.junit.Assert._
 import org.junit.Assume._
 
+import org.scalajs.testsuite.utils.AssertThrows.assertThrows
+
 /** Tests the little reflection we support */
 class ReflectionTest {
   import ReflectionTest._
@@ -131,6 +133,112 @@ class ReflectionTest {
     if (hiddenJSObj.getClass() != null)
       fail("optimizer thought that hiddenJSObj.getClass() was non-null")
   }
+
+  @Test def jsTypesKeptOnlyForTheirData_Issue4850(): Unit = {
+    /* Note: it is not possible to write `classOf[SomeObject.type]`. In order
+     * to get the class data of module classes (`object`s) without
+     * instantiating them or reaching anything else about them, we go through
+     * the `classOf` of an `Array[SomeObject.type]` then extract its
+     * `getComponentType()`.
+     */
+
+    import JSTypesKeptOnlyForTheirData._
+
+    @noinline
+    def nameOf(cls: Class[_]): String = cls.getName()
+
+    @inline
+    def testName(expectedShortName: String, cls: Class[_]): Unit = {
+      val prefix = "org.scalajs.testsuite.compiler.ReflectionTest$JSTypesKeptOnlyForTheirData$"
+      val expectedName = prefix + expectedShortName
+
+      assertEquals(expectedName, cls.getName()) // constant-folded
+      assertEquals(expectedName, nameOf(cls)) // evaluated at run-time
+    }
+
+    testName("NativeClass", classOf[NativeClass])
+    testName("NativeObject$", classOf[Array[NativeObject.type]].getComponentType())
+    testName("NativeTrait", classOf[NativeTrait])
+    testName("NonNativeClass", classOf[NonNativeClass])
+    testName("NonNativeObject$", classOf[Array[NonNativeObject.type]].getComponentType())
+    testName("NonNativeTrait", classOf[NonNativeTrait])
+
+    @noinline
+    def isInterfaceOf(cls: Class[_]): Boolean = cls.isInterface()
+
+    @inline
+    def testIsInterface(expected: Boolean, cls: Class[_]): Unit = {
+      assertEquals(expected, cls.isInterface()) // could be constant-folded in the future
+      assertEquals(expected, isInterfaceOf(cls)) // evaluated at run-time
+    }
+
+    // Consistent with isInterfaceForInstantiatedJSTypes()
+    testIsInterface(false, classOf[NativeClass])
+    testIsInterface(false, classOf[Array[NativeObject.type]].getComponentType())
+    testIsInterface(false, classOf[NativeTrait])
+    testIsInterface(false, classOf[NonNativeClass])
+    testIsInterface(false, classOf[Array[NonNativeObject.type]].getComponentType())
+    testIsInterface(false, classOf[NonNativeTrait])
+
+    @noinline
+    def isInstance(cls: Class[_], x: Any): Boolean = cls.isInstance(x)
+
+    @inline
+    def testIsInstance(expected: Boolean, cls: Class[_], x: Any): Unit = {
+      assertEquals(expected, cls.isInstance(x))
+      assertEquals(expected, isInstance(cls, x))
+    }
+
+    @inline
+    def testIsInstanceThrows(expected: js.Dynamic, cls: Class[_], x: Any): Unit = {
+      val e1 = assertThrows(classOf[js.JavaScriptException], cls.isInstance(x))
+      assertTrue(e1.toString(), js.special.instanceof(e1.exception, expected))
+
+      val e2 = assertThrows(classOf[js.JavaScriptException], isInstance(cls, x))
+      assertTrue(e2.toString(), js.special.instanceof(e2.exception, expected))
+    }
+
+    val jsDate: Any = new js.Date(1684246473882.0)
+
+    testIsInstance(false, classOf[NonNativeClass], jsDate)
+    testIsInstance(true, classOf[JSDateForIsInstance], jsDate)
+
+    // NativeClass is not actually defined, so isInstance will throw a ReferenceError
+    testIsInstanceThrows(js.constructorOf[js.ReferenceError], classOf[NativeClass], jsDate)
+
+    // isInstance is not supported on JS objects and traits; it throws a TypeError by spec
+    testIsInstanceThrows(js.constructorOf[js.TypeError],
+        classOf[Array[NativeObject.type]].getComponentType(), jsDate)
+    testIsInstanceThrows(js.constructorOf[js.TypeError], classOf[NativeTrait], jsDate)
+    testIsInstanceThrows(js.constructorOf[js.TypeError],
+        classOf[Array[NonNativeObject.type]].getComponentType(), jsDate)
+    testIsInstanceThrows(js.constructorOf[js.TypeError], classOf[NonNativeTrait], jsDate)
+    testIsInstanceThrows(js.constructorOf[js.TypeError],
+        classOf[Array[JSDateForIsInstance.type]].getComponentType(), jsDate)
+  }
+
+  @Test def isInterfaceForInstantiatedJSTypes(): Unit = {
+    // Make sure the instantiated non-native things are actually instantiated
+    assertEquals("function", js.typeOf(js.constructorOf[InstantiatedNonNativeClass]))
+    assertEquals("object", js.typeOf(InstantiatedNonNativeObject))
+
+    @noinline
+    def isInterfaceOf(cls: Class[_]): Boolean = cls.isInterface()
+
+    @inline
+    def testIsInterface(expected: Boolean, cls: Class[_]): Unit = {
+      assertEquals(expected, cls.isInterface()) // could be constant-folded in the future
+      assertEquals(expected, isInterfaceOf(cls)) // evaluated at run-time
+    }
+
+    // Consistent with jsTypesKeptOnlyForTheirData_Issue4850()
+    testIsInterface(false, classOf[js.Date]) // native class
+    testIsInterface(false, classOf[Array[js.Math.type]].getComponentType()) // native object
+    testIsInterface(false, classOf[js.Function0[Any]]) // native trait
+    testIsInterface(false, classOf[InstantiatedNonNativeClass])
+    testIsInterface(false, classOf[Array[InstantiatedNonNativeObject.type]].getComponentType())
+    testIsInterface(false, classOf[PseudoInstantiatedNonNativeTrait])
+  }
 }
 
 object ReflectionTest {
@@ -143,4 +251,36 @@ object ReflectionTest {
 
   class OtherPrefixRenamedTestClass
 
+  object JSTypesKeptOnlyForTheirData {
+    @js.native
+    @JSGlobal("NativeClass")
+    class NativeClass extends js.Object
+
+    @js.native
+    @JSGlobal("NativeObject")
+    object NativeObject extends js.Object
+
+    @js.native
+    trait NativeTrait extends js.Object
+
+    class NonNativeClass extends js.Object
+
+    object NonNativeObject extends js.Object
+
+    trait NonNativeTrait extends js.Object
+
+    @js.native
+    @JSGlobal("Date")
+    class JSDateForIsInstance extends js.Object
+
+    @js.native
+    @JSGlobal("Date")
+    object JSDateForIsInstance extends js.Object
+  }
+
+  trait PseudoInstantiatedNonNativeTrait extends js.Object
+
+  class InstantiatedNonNativeClass extends js.Object with PseudoInstantiatedNonNativeTrait
+
+  object InstantiatedNonNativeObject extends js.Object with PseudoInstantiatedNonNativeTrait
 }

